src/lib/parser.c

```diff
@@ -1678,8 +1682,103 @@ RdbStatus elementFunction(RdbParser *p) {
+RdbStatus elementModule(RdbParser *p) {
+    ElementCtx *ctx = &p->elmCtx;
+    enum MODULE_STATES {
+        ST_GET_MODULE_ID=0,
+
+        /* Following enums are aligned to module-opcodes to save mapping. Static assert below. */
+        ST_MODULE_OPCODE_SINT,
+        ST_MODULE_OPCODE_UINT,
+        ST_MODULE_OPCODE_FLOAT,
+        ST_MODULE_OPCODE_DOUBLE,
+        ST_MODULE_OPCODE_STRING,
+
+        ST_MODULE_NEXT_OPCODE,
+    } ;
+    while (1)
+    {
+        switch (ctx->state) {
+            case ST_GET_MODULE_ID: {
+                uint64_t moduleId; /*UNUSED*/
+                IF_NOT_OK_RETURN(rdbLoadLen(p, NULL, &moduleId, NULL, NULL));
+                /*** ENTER SAFE STATE ***/
+                updateElementState(p, ST_MODULE_NEXT_OPCODE);
+                break;
+            }
+            case ST_MODULE_OPCODE_SINT:
+            case ST_MODULE_OPCODE_UINT: {
+                uint64_t val; /*UNUSED*/
+                IF_NOT_OK_RETURN(rdbLoadLen(p, NULL, &val, NULL, NULL));
+                /*** ENTER SAFE STATE ***/
+                updateElementState(p, ST_MODULE_NEXT_OPCODE);
+                break;
+            }
+            case ST_MODULE_OPCODE_FLOAT: {
+                float val; /*UNUSED*/
+                IF_NOT_OK_RETURN(rdbLoadFloatValue(p, &val));
+                /*** ENTER SAFE STATE ***/
+                updateElementState(p, ST_MODULE_NEXT_OPCODE);
+                break;
+            }
+            case ST_MODULE_OPCODE_DOUBLE: {
+                double val; /*UNUSED*/
+                IF_NOT_OK_RETURN(rdbLoadDoubleValue(p, &val));
+                /*** ENTER SAFE STATE ***/
+                updateElementState(p, ST_MODULE_NEXT_OPCODE);
+                break;
+            }
+            case ST_MODULE_OPCODE_STRING: {
+                BulkInfo *bInfo; /*UNUSED*/
+                IF_NOT_OK_RETURN(rdbLoadString(p, RQ_ALLOC, NULL, &bInfo));
+                /*** ENTER SAFE STATE ***/
+                updateElementState(p, ST_MODULE_NEXT_OPCODE);
+                break;
+            }
+            case ST_MODULE_NEXT_OPCODE: {
+                IF_NOT_OK_RETURN(rdbLoadLen(p, NULL, &ctx->module.opcode, NULL, NULL));
+                /*** ENTER SAFE STATE ***/
+                if ((int)ctx->module.opcode == RDB_MODULE_OPCODE_EOF)
+                    return nextParsingElement(p, PE_END_KEY);
+
+                /* Cast Opcode to RDB_MODULE_OPCODE_*. Static assert below for mapping validation */
+                updateElementState(p, (int) ctx->module.opcode);
+                break;
+            }
+            default:
+                RDB_reportError(p, RDB_ERR_MODULE_INVALID_STATE,
+                    "elementModule() : Invalid parsing element state: %d.", ctx->state);
+                return RDB_STATUS_ERROR;
+        }
+    }
+
+    /* Static assert that verifies enums are aligned */
+    _Static_assert(ST_MODULE_OPCODE_SINT==RDB_MODULE_OPCODE_SINT, "enums must be aligned");
```

Comment:
`_Static_assert` was introduced in c11 if I'm not mistaken. gcc is generally doing terrible about warning these issues but clang would complain most probably.

Reply:
Let's refactor such kind of code when we goanna have CI that support clang to verify it.

Reply:
Just fyi, In my experience, compilers don't warn you reliably about the standard (it should be an easy thing to do though). Generally, you don't know until compile fails on some platform. So, I try to avoid features if I know it does not exist in a standard or on a platform. 

Anyway, here it is clang build, maybe it will throw some warnings: https://github.com/redis/librdb/pull/18

Reply:
As there is a lot of vudu code to have cross-version code for assert, maybe we can limit static-assert just for some of the versoins which is good enough.

Reply:
not sure I understand. Do you want to keep static assert? If so, this is how it is defined on my platform in cdefs.h: 

```c
#define _Static_assert(expr, diagnostic) \
    extern int (*__Static_assert_function (void)) \
      [!!sizeof (struct { int __error_if_negative: (expr) ? 2 : -1; })]
```

I guess you can define this if undefined. 


Btw, maybe you can assign those defines directly and avoid checking it altogether:

```c
enum {
    ST_MODULE_OPCODE_SINT = RDB_MODULE_OPCODE_SINT 
    ...
}
```

Reply:
i wanna remind us that these values are part of the rdb format and are never gonna change.
it's not like some other (volatile-only) enum, that can change but two components must remain allied.
it's like asserting that TCP is protocol number 6 in the IP packet header.
not needed IMHO.

Reply:
@tezc, 
I wrote it first like that but then I thought about the issue that enum doesn't verifies duplication of values.
But then thinking about it again, it might be good compromise between readability of the code and alignment enforcement. I think it I will go with your suggestion. 

@oranagra,
The verification is around changing the states values of the element that might break the alignment with the rdb opcodes. Not to verify that rdb opcodes has suddenly changed.